mesonbuild/build.py

```diff
@@ -1571,6 +1629,16 @@ def uses_rust(self) -> bool:
         elif self.generated:
             if self.generated[0].get_outputs()[0].endswith('.rs'):
                 return True
+        elif self.structured_sources:
+            for v in self.structured_sources.sources.values():
+                for s in v:
+                    if isinstance(s, (str, File)):
+                        if s.endswith('.rs'):
+                            return True
+                    else:
+                        for ss in s.get_outputs():
+                            if ss.endswith('.rs'):
```

Comment:
Should all the outputs end in .rs?

Reply:
No, this just checks that *any* file is a rust file. We have a sort of odd restriction that you can't mix rust and non-rust targets. Which doesn't really make sense. You can compile .c files to .o files, and link them with rust targets, we'd just need to generate a separate archiver step with Rust. It's on my list of things to do.